src/pybind/mgr/dashboard/services/exception.py

```diff
@@ -0,0 +1,136 @@
+    Or, as a replacement for cherrypy.HTTPError(...)
+    """
+
+    # pylint: disable=too-many-arguments
+    def __init__(self, e=None, code=None, component=None, http_status_code=None, msg=None):
+        super(DashboardException, self).__init__(msg)
+        self._code = code
+        self.component = component
+        if e:
+            self.e = e
+        if http_status_code:
+            self.status = http_status_code
+        else:
+            self.status = 400
+
+    def __str__(self):
+        try:
+            return str(self.e)
+        except AttributeError:
+            return super(DashboardException, self).__str__()
+
+    @property
+    def errno(self):
+        return self.e.errno
+
+    @property
+    def code(self):
+        if self._code:
+            return str(self._code)
+        return str(abs(self.errno))
+
+
+def serialize_dashboard_exception(e, include_http_status=False):
+    """
+    :type e: Exception
+    :param include_http_status: Used for Tasks, where the HTTP status code is not available.
+    """
+    from ..tools import ViewCache
+    if isinstance(e, ViewCacheNoDataException):
+        return {'status': ViewCache.VALUE_NONE, 'value': None}
+
+    out = dict(detail=str(e))
+    try:
+        out['errno'] = e.errno
+    except AttributeError:
+        pass
+    try:
+        out['code'] = e.code
+    except AttributeError:
+        pass
+    component = getattr(e, 'component', None)
+    out['component'] = component if component else None
+    if include_http_status:
+        out['status'] = getattr(e, 'status', 500)
+    return out
+
+
+def dashboard_exception_handler(handler, *args, **kwargs):
+    try:
+        with handle_rados_error(component=None):  # make the None controller the fallback.
+            return handler(*args, **kwargs)
+    # Don't catch cherrypy.* Exceptions.
+    except (ViewCacheNoDataException, DashboardException) as e:
+        logger.exception('dashboard_exception_handler')
+        cherrypy.response.headers['Content-Type'] = 'application/json'
+        cherrypy.response.status = getattr(e, 'status', 400)
+        return json.dumps(serialize_dashboard_exception(e)).encode('utf-8')
+
+
+@contextmanager
+def handle_rbd_error():
+    try:
+        yield
+    except rbd.OSError as e:
+        raise DashboardException(e, component='rbd')
+    except rbd.Error as e:
+        raise DashboardException(e, component='rbd', code=e.__class__.__name__)
+
+
+@contextmanager
+def handle_rados_error(component):
+    try:
+        yield
+    except rados.OSError as e:
+        raise DashboardException(e, component=component)
+    except rados.Error as e:
+        raise DashboardException(e, component=component, code=e.__class__.__name__)
+
+
+@contextmanager
+def handle_send_command_error(component):
+    try:
+        yield
+    except SendCommandError as e:
+        raise DashboardException(e, component=component)
+
+
+def c2d(my_contextmanager, *cargs, **ckwargs):
```

Comment:
can we rename this decorator to `ExceptionHandler`?

Reply:
My favorite would be to fix `@contentextmanager` to do the right thing. I'm thinking about:
```python
def correct_contextmanager(func):
    func = contextmanager(func):
    func.__call__ = c2d(func)
    return func
```
Then, it just works™